libvast/include/vast/pipeline.hpp

```diff
@@ -215,44 +218,66 @@ class operator_base {
     -> caf::expected<operator_type>;
 };
 
-/// A pipeline is a sequence of pipeline operators.
+/// A pipeline is a sequence of immutable operators with matching inputn and
+/// output element types.
 class pipeline final : public operator_base {
 public:
-  /// Constructs an empty pipeline.
+  /// Constructors, assignment operators, and a destructor.
   pipeline() = default;
-
+  ~pipeline() noexcept override = default;
   pipeline(pipeline const& other);
   pipeline(pipeline&& other) noexcept = default;
   auto operator=(pipeline const& other) -> pipeline&;
   auto operator=(pipeline&& other) noexcept -> pipeline& = default;
 
-  /// Constructs a pipeline from a sequence of operators. Flattens nested
-  /// pipelines, for example `(a | b) | c` becomes `a | b | c`.
-  explicit pipeline(std::vector<operator_ptr> operators);
+  /// Constructs a pipeline from a sequence of operators, which themselves may
+  /// be pipelines. If the definition is omitted, it is inferred from the
+  /// list of operators.
+  /// TODO: Make definition optional.
+  explicit pipeline(std::vector<operator_ptr> operators,
+                    std::optional<std::string> definition);
 
   /// Parses a pipeline from its definition.
   ///
   /// NOTE: If `location() != operator_location::local`, then
   /// `pipeline::parse(to_string(), {})` must succeed and be semantically
   /// equivalent to `*this`.
-  static auto parse(std::string_view repr) -> caf::expected<pipeline>;
+  static auto parse(std::string definition) -> caf::expected<pipeline>;
 
   /// @copydoc parse
-  static auto parse_as_operator(std::string_view repr)
+  static auto parse_as_operator(std::string definition)
     -> caf::expected<operator_ptr>;
 
   /// Adds an operator at the end of this pipeline.
-  void append(operator_ptr op);
+  void push_back(operator_ptr op);
 
   /// Adds an operator at the start of this pipeline.
-  void prepend(operator_ptr op);
+  void push_front(operator_ptr op);
 
   /// Returns the sequence of operators that this pipeline was built from.
-  auto unwrap() && -> std::vector<operator_ptr>;
+  auto unwrap() const& -> generator<std::pair<offset, const operator_base*>>;
+  auto unwrap() && -> generator<std::pair<offset, const operator_base*>>
+    = delete;
```

Comment:
Wdyt about making `operator_ptr` a `shared_ptr<const operator_base>`? That way, we could hand out shared pointers here.

Reply:
That would be one way to do it, although for most (all?) places where this would be used, `std::unique_ptr` suffices (as in the original `unwrap()` interface).